tests/pytests/test_search_params.py

```diff
@@ -0,0 +1,382 @@
+    env.assertEqual(conn.execute_command('HSET', 'key2', 'name', 'Alice', 'id', '31'), 2L)
+    env.assertEqual(conn.execute_command('HSET', 'key3', 'name', 'Carol', 'id', '13'), 2L)
+    env.assertEqual(conn.execute_command('HSET', 'key4', 'name', 'John\\ Doe', 'id', '0'), 2L)
+    env.assertEqual(conn.execute_command('HSET', 'key5', 'name', '$val1', 'id', '99'), 2L)
+    env.assertEqual(conn.execute_command('HSET', 'key6', 'name', 'John Doh', 'id', '100'), 2L)
+    env.assertEqual(conn.execute_command('HSET', 'key7', 'name', 'John', 'id', '100'), 2L)
+
+    # Test expression
+    res1 = conn.execute_command('FT.SEARCH', 'idx', '@name:(Alice|Bob)', 'NOCONTENT')
+    env.assertEqual(res1, [2L, 'key2', 'key1'])
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '@name:($val1|Bob)', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Alice')
+    env.assertEqual(res2, res1)
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '@name:(Alice|$val1)', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Bob')
+    env.assertEqual(res2, res1)
+
+    res1 = conn.execute_command('FT.SEARCH', 'idx', '@name:(Alice)', 'NOCONTENT')
+    env.assertEqual(res1, [1L, 'key2'])
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '@name:($val1)', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Alice')
+    env.assertEqual(res2, res1)
+
+    res1 = conn.execute_command('FT.SEARCH', 'idx', '@name:(John\\ Doe)', 'NOCONTENT')
+    env.assertEqual(res1, [1L, 'key4'])
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '@name:($val1)', 'NOCONTENT', 'PARAMS', '2', 'val1', 'John\\ Doe')
+    env.assertEqual(res2, res1)
+
+    # Test negative expression
+    res1 = conn.execute_command('FT.SEARCH', 'idx', '-(@name:(Alice|Bob))', 'NOCONTENT')
+    env.assertEqual(res1, [5L, 'key3', 'key4', 'key5', 'key6', 'key7'])
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '-(@name:($val1|Bob))', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Alice')
+    env.assertEqual(res2, res1)
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '-(@name:($val1|Bob))', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Alice')
+    env.assertEqual(res2, res1)
+
+    # Test optional token
+    res1 = conn.execute_command('FT.SEARCH', 'idx', '@name:(John ~Doh)', 'NOCONTENT')
+    env.assertEqual(res1, [2L, 'key6', 'key7'])
+    res2 = conn.execute_command('FT.SEARCH', 'idx', '@name:(John ~$val1)', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Doh')
+    env.assertEqual(res2, res1)
+
+    # FIXME: Avoid parameterization in verbatim string (whether a param is defined or not)
+    #  Parser seems OK
+    #  (need to review indexing, in previous versions the following search query was syntactically illegal)
+    # res1 = conn.execute_command('FT.SEARCH', 'idx', '@name:("$val1")', 'NOCONTENT')
+    # env.assertEqual(res1, [1L, 'key5'])
+    # res2 = conn.execute_command('FT.SEARCH', 'idx', '@name:("$val1")', 'NOCONTENT', 'PARAMS', '2', 'val1', 'Alice')
+    # env.assertEqual(res2, res1)
```

Comment:
what about this one?

Reply:
@rafie This is a pending completion, probably on indexing (in Search 2.0 a `$` here was causing a syntax error)